File: Материалы для проверки решений рег этап 2015/Пробный тур/Задача Z Конфеты/sweets_sk_cheat.cpp
/**
 * Author: Sergey Kopeliovich (Burunduk30@gmail.com)
 * Date: 2013.11.15
 *
 * Comment: O(1), try only +-8
 */

#include <cstdio>
#include <cassert>
#include <iostream>
#include <algorithm>

using namespace std;

#define forn(i, n) for (int i = 0; i < (int)(n); i++)
#define forab(i, a, b) for (int i = (a); i <= (b); i++)

typedef long long ll;
typedef long double dbl;

int main()
{
  #define NAME "sweets"
  assert(freopen(NAME ".in", "r", stdin));
  assert(freopen(NAME ".out", "w", stdout));

  int n, a, b, c;
  assert(scanf("%d%d%d%d", &n, &a, &b, &c) == 4);

  const int M = 8;
  int x0 = n / (3 * a);
  int y0 = n / (3 * b);
  int z0 = n / (3 * c);
  int rx = x0, ry = y0, rz = z0;
  dbl best = (ll)x0 * y0 * z0, tmp;
  forab(dx, -M, M)
    forab(dy, -M, M)
      forab(dz, -M, M)
      {
        int x = x0 + dx, y = y0 + dy, z = z0 + dz;
        if (1 <= x && 1 <= y && 1 <= z && (ll)a * x + (ll)b * y + (ll)c * z <= n)
          if ((tmp = (dbl)x * y * z) - best > 0)
            best = tmp, rx = x, ry = y, rz = z;
      }
  
  printf("%d %d %d\n", a * rx, b * ry, c * rz);
  return 0;
}

File: Материалы для проверки решений рег этап 2015/Пробный тур/Задача Z Конфеты/sweets_sk_cheat.py
#
# Author: Sergey Kopeliovich (Burunduk30@gmail.com)
# Date: 2013.11.15
#
# Comment: O(1), try only +-8, no overflow
#

import sys

NAME = "sweets"

inf = open(NAME + ".in", "r")
out = open(NAME + ".out", "w")

n, a, b, c = list(map(int, inf.readline().split()))

x0 = n // (3 * a)
y0 = n // (3 * b)
z0 = n // (3 * c)
best = x0 * y0 * z0
rx, ry, rz = x0, y0, z0
M = 8
r = range(-M, M + 1)
for dx in r: 
  for dy in r:
    for dz in r:
      x, y, z = x0 + dx, y0 + dy, z0 + dz
      if ((1 <= x) and (1 <= y) and (1 <= z) and (a * x + b * y + c * z <= n)):
        tmp = x * y * z
        if (tmp > best):
          best, rx, ry, rz = tmp, x, y, z

print(rx * a, ry * b, rz * c, file=out)

inf.close()
out.close()
